File: backend/src/main/java/com/poeflipfinder/backend/framework/divinationcard/BundledDivinationCardReferenceGateway.java
package com.poeflipfinder.backend.framework.divinationcard;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.poeflipfinder.backend.entity.Currency;
import com.poeflipfinder.backend.entity.DivinationCardReward;
import com.poeflipfinder.backend.gateway.DivinationCardReferenceGateway;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.List;
import org.springframework.core.io.ClassPathResource;

/**
 * The Frameworks & Drivers implementation of DivinationCardReferenceGateway,
 * reading the manually-captured divination card reward data (docs/DATA_SOURCES.md
 * § Divination Card Turn-In Rewards) from a bundled resource -- same treatment
 * as {@link com.poeflipfinder.backend.framework.itemicon.GggItemIconGateway}'s
 * Item Icons catalog, since neither source has a live API to fetch from.
 *
 * <p>The {@code card}/{@code rewardCurrency} fields of the returned
 * {@link DivinationCardReward} entities are name-only placeholders (no id,
 * externalId, or iconUrl) -- this gateway has no access to live Currency
 * Exchange data to resolve those against. Callers must match by
 * {@link Currency#displayName()} against currently-active market snapshots to
 * get a fully-resolved Currency before using one in a FlipOpportunity.
 */
public class BundledDivinationCardReferenceGateway implements DivinationCardReferenceGateway {

    private static final String CATALOG_RESOURCE_PATH = "reference-data/divination-card-rewards.json";

    private final ObjectMapper objectMapper;
    private List<DivinationCardReward> cardRewards;

    public BundledDivinationCardReferenceGateway() {
        this.objectMapper = new ObjectMapper();
    }

    @Override
    public List<DivinationCardReward> findAll() {
        ensureCatalogLoaded();
        return cardRewards;
    }

    private synchronized void ensureCatalogLoaded() {
        if (cardRewards != null) {
            return;
        }
        normalize(readBundledCatalog());
    }

    private String readBundledCatalog() {
        try (InputStream in = new ClassPathResource(CATALOG_RESOURCE_PATH).getInputStream()) {
            return new String(in.readAllBytes(), StandardCharsets.UTF_8);
        } catch (IOException e) {
            throw new DivinationCardReferenceGatewayException(
                    "Failed to read bundled divination card catalog " + CATALOG_RESOURCE_PATH, e);
        }
    }

    /** Package-visible so a test can exercise parsing without touching the classpath resource. */
    void normalize(String json) {
        List<RawEntry> entries = deserialize(json);
        cardRewards = entries.stream().map(this::toDomain).toList();
    }

    private List<RawEntry> deserialize(String json) {
        try {
            return objectMapper.readValue(json, objectMapper.getTypeFactory().constructCollectionType(List.class, RawEntry.class));
        } catch (IOException e) {
            throw new DivinationCardReferenceGatewayException("Failed to parse divination card catalog", e);
        }
    }

    private DivinationCardReward toDomain(RawEntry entry) {
        Currency card = new Currency(null, null, entry.cardName(), null, Currency.ItemType.DIVINATION_CARD);
        Currency rewardCurrency = entry.rewardCurrencyName() == null
                ? null
                : new Currency(null, null, entry.rewardCurrencyName(), null, Currency.ItemType.CURRENCY);
        int rewardQuantity = entry.rewardQuantity() == null ? 0 : entry.rewardQuantity();
        return new DivinationCardReward(card, entry.stackSize(), rewardCurrency, rewardQuantity, entry.isPredictable());
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private record RawEntry(
            String cardName, int stackSize, String rewardCurrencyName, Integer rewardQuantity, boolean isPredictable) {
    }
}

File: app/lib/poe_flip_finder/gateways/bundled_divination_card_reference_gateway.ex
defmodule PoeFlipFinder.Gateways.BundledDivinationCardReferenceGateway do
  @moduledoc """
  The imperative-shell implementation of
  `PoeFlipFinder.DivinationCardReferenceGateway`, reading manually-captured
  PoE Wiki turn-in reward data (docs/DATA_SOURCES.md § Divination Card
  Turn-In Rewards) from a bundled resource -- same treatment as
  `GggItemIconGateway`'s Item Icons catalog, since neither source has a
  live API to fetch from.

  The `card`/`reward_currency` fields of returned `DivinationCardReward`
  entities are name-only placeholders (no `id`, `external_id`, or
  `icon_url`) -- this gateway has no access to live Currency Exchange data
  to resolve those against. Callers must match by `Currency.display_name`
  against currently-active market snapshots to get a fully-resolved
  Currency before using one in a FlipOpportunity.

  The parsed list is cached in `:persistent_term` after first load, same
  rationale as `GggItemIconGateway`: small, effectively-immutable data read
  on every flip-opportunities computation.
  """

  @behaviour PoeFlipFinder.DivinationCardReferenceGateway

  alias PoeFlipFinder.{Currency, DivinationCardReward}

  @catalog_resource_path "reference-data/divination-card-rewards.json"
  @persistent_term_key {__MODULE__, :card_rewards}

  @impl true
  def find_all, do: ensure_loaded()

  defp ensure_loaded do
    case :persistent_term.get(@persistent_term_key, nil) do
      nil ->
        rewards = normalize(Jason.decode!(read_bundled_catalog()))
        :persistent_term.put(@persistent_term_key, rewards)
        rewards

      rewards ->
        rewards
    end
  end

  defp read_bundled_catalog do
    :poe_flip_finder
    |> Application.app_dir("priv")
    |> Path.join(@catalog_resource_path)
    |> File.read!()
  end

  @doc "Exposed so contract tests can exercise parsing without reading the bundled file."
  @spec normalize([map()]) :: [DivinationCardReward.t()]
  def normalize(raw_entries), do: Enum.map(raw_entries, &to_entity/1)

  defp to_entity(entry) do
    %DivinationCardReward{
      card: placeholder_currency(entry["cardName"], :cards),
      stack_size: entry["stackSize"],
      reward_currency: placeholder_reward_currency(entry["rewardCurrencyName"]),
      reward_quantity: entry["rewardQuantity"] || 0,
      predictable: entry["isPredictable"]
    }
  end

  defp placeholder_reward_currency(nil), do: nil
  defp placeholder_reward_currency(name), do: placeholder_currency(name, :currency)

  defp placeholder_currency(name, category) do
    %Currency{
      id: nil,
      external_id: nil,
      display_name: name,
      icon_url: nil,
      category: category
    }
  end
end
